src/routes/v3/oauth_clients.rs

```diff
@@ -0,0 +1,387 @@
+use std::{collections::HashSet, fmt::Display};
+
+use actix_web::{
+    delete, get, patch, post,
+    web::{self},
+    HttpRequest, HttpResponse,
+};
+use chrono::Utc;
+use itertools::Itertools;
+use rand::{distributions::Alphanumeric, Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use serde::{Deserialize, Serialize};
+use sqlx::PgPool;
+use validator::Validate;
+
+use super::ApiError;
+use crate::{
+    auth::get_user_from_headers,
+    database::{
+        models::{
+            generate_oauth_client_id, generate_oauth_redirect_id,
+            oauth_client_authorization_item::OAuthClientAuthorization,
+            oauth_client_item::{OAuthClient, OAuthRedirectUri},
+            DatabaseError, OAuthClientId, User,
+        },
+        redis::RedisPool,
+    },
+    models::{self, oauth_clients::OAuthClientCreationResult, pats::Scopes},
+    queue::session::AuthQueue,
+    routes::v2::project_creation::CreateError,
+    util::validate::validation_errors_to_string,
+};
+
+use crate::database::models::oauth_client_item::OAuthClient as DBOAuthClient;
+
+pub fn config(cfg: &mut web::ServiceConfig) {
+    cfg.service(oauth_client_create);
+    cfg.service(oauth_client_edit);
+    cfg.service(oauth_client_delete);
+    cfg.service(get_user_clients);
+    cfg.service(get_user_oauth_authorizations);
+    cfg.service(revoke_oauth_authorization);
+}
```

Comment:
i think a better architecture would be having independent routes for retrieving oauth clients by ids, like single and multiple. and then other routes can just refer to the ids

Reply:
Sure, will do

Reply:
Alright, reworked the routes structure to the following:    
    Get user clients: GET /user/{user_id}/oauth_apps
        - Gets the clients for the current user based on header
        - Returns vec of client ids, no details
    Get client(s): GET /oauth/apps
	    - With json body { ids=[{id_1}, {id_2}, ...] }
    Get client: GET /oauth/app/{client_id}
    Create client: POST /oauth/app
    Edit client: PATCH /oauth/app/{client_id}
    Delete client: DELETE /oauth/app/{client_id}
    Get current user authorizations: GET /oauth/authorizations
        - Returns vec of user's authorizations with detail about authorization, but only client id, no other client details
    Revoke current user authorization: DELETE /oauth/authorizations?client_id={client_id}